frontend/src/metabase/api/tags/utils.ts

```diff
@@ -0,0 +1,183 @@
+import type { TagDescription } from "@reduxjs/toolkit/query";
+
+import type {
+  Card,
+  Collection,
+  Database,
+  Field,
+  FieldDimension,
+  FieldId,
+  ForeignKey,
+  Metric,
+  Segment,
+  Table,
+  Timeline,
+  TimelineEvent,
+  UserInfo,
+} from "metabase-types/api";
+
+import type { TagType } from "./constants";
+
+export function tag(type: TagType): TagDescription<TagType> {
+  return { type };
+}
+
+export function listTag(type: TagType): TagDescription<TagType> {
+  return { type, id: "LIST" };
+}
+
+export function idTag(
+  type: TagType,
+  id: string | number,
+): TagDescription<TagType> {
+  return { type, id };
+}
+
+export function invalidateTags(
+  error: unknown,
+  tags: TagDescription<TagType>[],
+): TagDescription<TagType>[] {
+  return !error ? tags : [];
+}
+
+export function databaseTags(database: Database): TagDescription<TagType>[] {
```

Comment:
Only staff that is used in multiple places is added here

Reply:
Each method also uses another, so they should be in one file

Reply:
Agree, on in

Reply:
@kamilmielnik fixed, please take a look. Now for all `providesTags` there is a function in `tags` folder. However invalidation is still done manually. So far it seems to be endpoint-specific and not something we can guess without the context